src/main/java/org/spongepowered/common/event/tracking/context/BlockPosMultiMap.java

```diff
@@ -0,0 +1,268 @@
+        if (!added) {
+            // Ok, means we have a multi change on a same position, now to use the multimap
+            // for probably the first time.
+            if (this.multimap == null) {
+                // If the multimap hasn't been populated yet, well, now we have to populate the multimap and eliminate the list.
+                this.multimap = LinkedListMultimap.create(); // LinkedListMultimap is insertion order respective, so the backed lists per
+                // Now to populate it from the previously used list of snapshots...
+                for (SpongeBlockSnapshot existing : this.snapshots) { // Ignore snapshots potentially being null, it will never be null at this point.
+                    this.multimap.put(existing.getBlockPos(), existing);
+                }
+            }
+            // And place the snapshot into the multimap.
+            this.multimap.put(blockPos, backingSnapshot);
+            // Now we can re-evaluate the modified block position
+            // If the position is duplicated, we need to update the original snapshot of the now incoming block change
+            // in relation to the original state (so if a block was set to air, then afterwards set to piston head, it should go from break to modify)
+            final List<SpongeBlockSnapshot> list = this.multimap.get(blockPos);
+            if (list != null && !list.isEmpty()) {
+                final SpongeBlockSnapshot originalSnapshot = list.get(0);
+                TrackingUtil.associateBlockChangeWithSnapshot(PhaseTracker.getInstance().getCurrentState(), newState.getBlock(), BlockUtil.toNative(originalSnapshot.getState()), originalSnapshot );
+            }
+            // Flip the boolean to have fasts for next entry
+            this.hasMulti = true;
+            return false;
+        }
+        // At this point, we haven't captured the block position yet.
+        // and we can check if the list is null.
+        if (this.snapshots == null) {
+            this.snapshots = new ArrayList<>();
+        }
+        this.snapshots.add(backingSnapshot);
+        // And this is the only time that we return true, if we have not caught multiple transactions per position before.
+        return true;
+    }
+
+    public boolean hasMultiChanges() {
+        return this.hasMulti;
+    }
+
+    /**
+     * Gets a copied {@link List} of the first {@link BlockSnapshot}s for each
+     * {@link BlockPos} changed. The nature of this is that each individual
+     * {@link BlockPos} being changed will have an "original" snapshot to compare
+     * to what currently exists within a {@link World}. This list is intended not
+     * to be added to and will result in unmodifiable exceptions.
+     * @return
+     */
+    public final List<SpongeBlockSnapshot> get() {
+        return Collections.unmodifiableList(this.snapshots == null ? Collections.emptyList() : this.snapshots);
+    }
+
+    public final void prune(BlockSnapshot snapshot) {
```

Comment:
Should there be a method body, or can this be removed?

Reply:
guess this is still wip.